### PR TITLE
API consistency for System.Memory APIs (argument names)

### PR DESCRIPTION
**Pull request by ahsonkhan**
Here is an attempt to keep argument names consistent across the board for all the System.Memory APIs, based on our discussion (see https://github.com/dotnet/apireviews/pull/62).

I have split each change into separate commits if that makes things easier to review.

**List of changes:**

- AsBytes and Cast: Change arg name source -> span
- CopyTo: Change arg name array/sequence -> source
- Several APIs: Change arg name arraySegment -> segment
- Several APIs: Change arg name readOnlyMemory -> memory
- BinaryPrimitives and Utf8Formatter: Change arg name buffer -> source or destination
- Utf8Parser: Change arg name text -> source
- Base64: Change arg name consumed/written -> bytesConsumed/bytesWritten
- TryGetOwnedMemory: Change arg name ownedMemory -> owner, index -> start
- BuffersExtensions.Write: Change arg name bufferWriter -> writer
- Span Compare APIs: Change arg name first -> span, second/value -> other
- ROSequence.TryGet & BuffersExtensions: Change arg name data -> memory, sequence -> source
- BufferExtensions.Write: Change arg name source -> span
- SequencePosition.Equals: Change arg name position -> other

cc @KrzysztofCwalina, @joshfree, @terrajobst, @stephentoub, @davidfowl, @pakrym, @benaadams, @jkotas, @JeremyKuhne, @dotnet/corefxlab-contrib, @karelz, @danmosemsft

### CODE REVIEW

File: src/Common/src/CoreLib/System/MemoryExtensions.Fast.cs [hunk 1]
```diff
@@ -17,46 +17,46 @@ namespace System
...
-        /// Returns a value indicating whether the specified <paramref name="value"/> occurs within the <paramref name="span"/>.
+        /// Returns a value indicating whether the specified <paramref name="other"/> occurs within the <paramref name="span"/>.
         /// <param name="span">The source span.</param>
-        /// <param name="value">The value to seek within the source span.</param>
-        /// <param name="comparisonType">One of the enumeration values that determines how the <paramref name="span"/> and <paramref name="value"/> are compared.</param>
+        /// <param name="other">The value to seek within the source span.</param>
+        /// <param name="comparisonType">One of the enumeration values that determines how the <paramref name="span"/> and <paramref name="other"/> are compared.</param>
         /// </summary>
-        public static bool Contains(this ReadOnlySpan<char> span, ReadOnlySpan<char> value, StringComparison comparisonType)
+        public static bool Contains(this ReadOnlySpan<char> span, ReadOnlySpan<char> other, StringComparison comparisonType)
```
Comment (KrzysztofCwalina) [inline src/Common/src/CoreLib/System/MemoryExtensions.Fast.cs]: I think the "other" terminology is used only in Equals/Compare. I think in Contains "value" works better.

@terrajobst, what was the intention here?

File: src/System.Memory/ref/System.Memory.cs [hunk 2]
```diff
@@ -288,14 +288,14 @@ public abstract partial class OwnedMemory<T> : System.Buffers.IRetainable, Syste
...
     }
     public static partial class BuffersExtensions
     {
-        public static void CopyTo<T>(in this System.Buffers.ReadOnlySequence<T> sequence, System.Span<T> destination) { }
-        public static System.Nullable<System.SequencePosition> PositionOf<T>(in this System.Buffers.ReadOnlySequence<T> sequence, T value) where T : System.IEquatable<T> { throw null; }
-        public static T[] ToArray<T>(in this System.Buffers.ReadOnlySequence<T> sequence) { throw null; }
-        public static void Write<T>(this System.Buffers.IBufferWriter<T> bufferWriter, ReadOnlySpan<T> source) { }
+        public static void CopyTo<T>(in this System.Buffers.ReadOnlySequence<T> source, System.Span<T> destination) { }
+        public static System.Nullable<System.SequencePosition> PositionOf<T>(in this System.Buffers.ReadOnlySequence<T> source, T value) where T : System.IEquatable<T> { throw null; }
+        public static T[] ToArray<T>(in this System.Buffers.ReadOnlySequence<T> source) { throw null; }
```
Comment (KrzysztofCwalina) [inline src/System.Memory/ref/System.Memory.cs]: I think this parameter should be called "sequence". Just like AsBytes parameter is called "span".

@terrajobst?

File: src/System.Memory/ref/System.Memory.cs [hunk 3]
```diff
@@ -288,14 +288,14 @@ public abstract partial class OwnedMemory<T> : System.Buffers.IRetainable, Syste
...
+        public static void Write<T>(this System.Buffers.IBufferWriter<T> writer, ReadOnlySpan<T> span) { }
```
Comment (KrzysztofCwalina) [inline src/System.Memory/ref/System.Memory.cs]: Should this be called "value"?

@terrajobst?

**Comment by joshfree** [comment]
Thanks @ahsonkhan; let's never do this again. 😄

File: src/Common/src/CoreLib/System/MemoryExtensions.Fast.cs (hunk 1 above)
Comment (ahsonkhan) [reply src/Common/src/CoreLib/System/MemoryExtensions.Fast.cs]: What about the other APIs that were part of this change: StartsWith/EndsWith/Overlaps?

File: src/System.Memory/ref/System.Memory.cs (hunk 2 above)
Comment (ahsonkhan) [reply src/System.Memory/ref/System.Memory.cs]: I would be fine with that.

**Comment by ahsonkhan** [comment]
> Given that this code has lots of ifdefs / partial files shared with corelib, if you haven't already I'd suggest manually copying the relevant files to your coreclr repo and doing a build to ensure the renaming was applied to all occurrences.

Locally:
- build.cmd -allconfigurations completes successfully
- Copied shared files into coreclr, it also builds successfully

**Comment by ahsonkhan** [comment]
@dotnet-bot test Alpine.3.6 x64 Debug Build
@dotnet-bot test Windows x64 Debug Build

**Comment by ahsonkhan** [comment]
The dumpling failure is the same one as stated here - https://github.com/dotnet/corefx/pull/28225#issuecomment-374409319
cc @danmosemsft, @Anipik 

https://ci3.dot.net/job/dotnet_corefx/job/master/job/osx-TGroup_netcoreapp+CGroup_Debug+AGroup_x64+TestOuter_false_prtest/10006/console
```text
17:39:54   Traceback (most recent call last):
17:39:54     File "/Users/dotnet-bot/j/workspace/dotnet_corefx/master/osx-TGroup_netcoreapp+CGroup_Debug+AGroup_x64+TestOuter_false_prtest/Tools/DumplingHelper.py", line 131, in <module>
17:39:54       main(sys.argv)
17:39:54     File "/Users/dotnet-bot/j/workspace/dotnet_corefx/master/osx-TGroup_netcoreapp+CGroup_Debug+AGroup_x64+TestOuter_false_prtest/Tools/DumplingHelper.py", line 113, in main
17:39:54       install_dumpling()
17:39:54     File "/Users/dotnet-bot/j/workspace/dotnet_corefx/master/osx-TGroup_netcoreapp+CGroup_Debug+AGroup_x64+TestOuter_false_prtest/Tools/DumplingHelper.py", line 34, in install_dumpling
17:39:54       print("An unexpected error was encountered while installing dumpling.py: " + sys.exc_info()[0])
17:39:54   TypeError: cannot concatenate 'str' and 'type' objects
17:39:54 /Users/dotnet-bot/j/workspace/dotnet_corefx/master/osx-TGroup_netcoreapp+CGroup_Debug+AGroup_x64+TestOuter_false_prtest/Tools/Dumpling.targets(30,5): error MSB3073: The command "python /Users/dotnet-bot/j/workspace/dotnet_corefx/master/osx-TGroup_netcoreapp+CGroup_Debug+AGroup_x64+TestOuter_false_prtest/Tools/DumplingHelper.py install_dumpling" exited with code 1. [/Users/dotnet-bot/j/workspace/dotnet_corefx/master/osx-TGroup_netcoreapp+CGroup_Debug+AGroup_x64+TestOuter_false_prtest/src/tests.builds]
```

@dotnet-bot test Alpine.3.6 x64 Debug Build
@dotnet-bot test OSX x64 Debug Build

**Comment by Anipik** [comment]
@ahsonkhan PR https://github.com/dotnet/buildtools/pull/1969 has been opened to correct the issue

**Pull request merged**